### ISSUE TITLE
How many epochs did you train for Post Training before put into finetuning?

### ISSUE THREAD

**Issue by RayXu14** [taesunwhang/BERT-ResSel#4]
I suppose if you really train 3 epochs for 10 duplication data, it will take more than 10 days

**Comment by taesunwhang** [comment]
Hi, we set the number of epochs as 2 for post-training with 10 duplication data and the batch size was set to 384 (4 V100 GPUs). It took about 2 days for training.

**Comment by RayXu14** [comment]
Thank you for your reply! Do you mean virtual batch size = 384?

**Comment by taesunwhang** [comment]
No problem. we set actual batch to 384 by using multi-gpus (96 batch size for each gpu), but yes. If you are gonna train the model by using single gpu, set virtual batch size to 384.

**Comment by RayXu14** [comment]
In the bert-post-uncased-pytorch_model.pth you provided:
...
_bert_model.cls.predictions.bias
_bert_model.cls.predictions.**new_bias**
_bert_model.cls.predictions.transform.dense.weight
_bert_model.cls.predictions.transform.dense.bias
_bert_model.cls.predictions.transform.LayerNorm.weight
_bert_model.cls.predictions.transform.LayerNorm.bias
_bert_model.cls.predictions.decoder.weight
_bert_model.cls.seq_relationship.weight
_bert_model.cls.seq_relationship.bias 

In the origin bert-base-uncased bin file:
...
cls.predictions.bias
cls.predictions.transform.dense.weight
cls.predictions.transform.dense.bias
cls.predictions.transform.LayerNorm.gamma
cls.predictions.transform.LayerNorm.beta
cls.predictions.decoder.weight
cls.seq_relationship.weight
cls.seq_relationship.bias

There are differences in your saved state_dict and origin BERT's BertForPreTraining class, and in your pth file:
state_dict['model']['_bert_model.cls.predictions.decoder.weight'].shape = torch.Size([30523, 768]) 
state_dict['model']['_bert_model.cls.predictions.bias'].shape = torch.Size([30522])

but in BERT's pretrained bin file,
state_dict['cls.predictions.decoder.weight'].shape = torch.Size([30522, 768])

You write a new_bias, but do not modify the shape of cls.predictions.decoder.weight, how do the shape of weight of decoder change from 30522 to 30523?

In your bert_post_training.py, you load pretrained file from modeling_bert.BertModel, and does not create any pretraining prediction head, this means post training in the code you provided cannot be run.
I wonder if you make use of cls head in the origin BERT's pretrained weight or just initialize a new one?

**Comment by taesunwhang** [comment]
I added _bert_model.cls.predictions.**new_bias** for the special token **[EOT]**. 
For the case of decoder shape (state_dict['model']['_bert_model.cls.predictions.decoder.weight'].shape = torch.Size([30523, 768]))
, the token_embeddings is resized from 30522 to 30523 when building the BERT model. 
I found some bugs in **bert_post_training.py**, so I fixed it! Please use the updated code.

**Comment by RayXu14** [comment]
Let me make it more clear:

(decoder): Linear(in_features=768, out_features=30522, bias=False)

the weight of Linear layer in the decoder of MLM head is not changed to 30523 in the code

**Comment by RayXu14** [comment]
would you fix the first 8 layers in BERT when do post training?

**Comment by RayXu14** [comment]
I see you change the dupe_factor from 10 to 1, what is the intent of this modification?

**Comment by taesunwhang** [comment]
1. Let me make it more clear:
(decoder): Linear(in_features=768, out_features=30522, bias=False)
the weight of Linear layer in the decoder of MLM head is not changed to 30523 in the code
-> I just checked word embedding layer of post-trained BERT is [30523,768]. This embedding is will be shared for MLM.

2. would you fix the first 8 layers in BERT when do post training?
-> I didn't change the BERT layer during the post-training.

3. I see you change the dupe_factor from 10 to 1, what is the intent of this modification?

-> I just checked my code works properly. Set dupe_factor as 10 (no intent at all).

**Comment by RayXu14** [comment]
I appreciate that you spent your time to fix the code~ Please allow me to ask final 2 questions:

1. "This embedding is will be shared for MLM."
  -> You mean you use the transpose of the word embedding matrix as the weight in the decoder?
2. I mean freeze first 8 layers, not change the structure. To be clear, I am asking if you calculate the gradient of first 8 layers in BERT for post training or freeze it just like BERT-VFT?

**Comment by taesunwhang** [comment]
1. Yes. see **models/bert/modeling_bert.py** line **673**. 
https://github.com/taesunwhang/BERT-ResSel/blob/master/models/bert/modeling_bert.py
By using **tie_weights** method, we are sharing cls.predictions.decoder and bert.embeddings.word_embeddings.

2. In the case of BERT-VFT, we freeze the first 8 layers during the fine-tuning period (after post-training BERT). 

Hope my answers are helpful for your ongoing research. :)

**Comment by RayXu14** [comment]
Sorry to bother you again > _ <  Did the post training settings（2 epochs with 10 duplication data） of Advising dataset the same as Ubuntu?
